aodndata/moorings/products_handler.py

```diff
@@ -0,0 +1,98 @@
+
+    def __init__(self, *args, **kwargs):
+        super(MooringsProductsHandler, self).__init__(*args, **kwargs)
+        if self.allowed_extensions is None:
+            self.allowed_extensions = ['.json']
+        self.product_site_code = None
+        self.product_variables = None
+
+    def preprocess(self):
+        """Collect available input files and create the products, adding them to the collection to be published."""
+
+        # Don't want to publish the manifest file, so remove it from the collection
+        self.file_collection.pop()
+
+        # Read the manifest file and extract key parameters for product
+        with open(self.input_file) as f:
+            try:
+                manifest = json.load(f)
+            except ValueError as e:
+                raise InvalidFileContentError("invalid JSON file '{self.input_file}' ({e})".format(self=self, e=e))
+
+        try:
+            self.product_site_code = manifest['site_code']
+            self.product_variables = manifest['variables']
+        except KeyError:
+            raise InvalidFileContentError(
+                "manifest file '{self.input_file}' missing information (site_code, variables)".format(self=self)
+            )
+        self.logger.info(
+            "Creating products for site {self.product_site_code}, variables {self.product_variables}".format(self=self)
+        )
+
+        # Find out what relevant input files are available on S3 for this site.
+        filter_list = [PropertyIsEqualTo(propertyname='site_code', literal=self.product_site_code),
+                       PropertyIsEqualTo(propertyname='file_version', literal='1'),
+                       PropertyIsEqualTo(propertyname='realtime', literal='false'),
+                       PropertyIsNotEqualTo(propertyname='data_category', literal='Biogeochem_profiles'),
+                       PropertyIsNotEqualTo(propertyname='data_category', literal='CTD_profiles')
+                       ]
+        filter = ogc_filter_to_string(And(filter_list))
+
+        # Note I need to access _wfs_broker to be able to use query_urls_for_layer() with a filter,
+        # as the corresponding StateQuery method doesn't accept additional kwargs.
+        # TODO: replace ._wfs_broker.query_urls_for_layer() with .query_wfs_urls_for_layer() once aodncore has been updated
+        input_files = self.state_query._wfs_broker.query_urls_for_layer(self.FILE_INDEX_LAYER,
+                                                                        ogc_filter=filter,
+                                                                        url_property_name='url'
+                                                                        )
+        self.logger.info("Downloading {n} input files".format(n=len(input_files)))
+
+        # Download input files to local cache.
+        input_file_collection = RemotePipelineFileCollection(input_files)
+        input_file_collection.download(self._upload_store_runner.broker, self.temp_dir)
```

Comment:
We need to consider a standard way to do this, so that it doesn't need to access the private `_upload_store_runner` property. Perhaps by adding method to HandlerBase, which abstracts this detail from child handlers.

e.g.

`self.download_collection(input_file_collection)`

Reply:
Yep, sounds good.